File: services/drakon-agent/validator/ir_validator.py
"""Validates DRAKON IR structure before returning to caller."""
from dataclasses import dataclass, field


@dataclass
class ValidationResult:
    valid: bool
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


def validate_ir(ir: dict) -> ValidationResult:
    """Validate a single DRAKON IR diagram dict."""
    errors: list[str] = []
    warnings: list[str] = []

    if not isinstance(ir, dict):
        return ValidationResult(False, ["IR must be a dict"])

    # Required fields
    if "name" not in ir:
        errors.append("Missing 'name' field")
    if "items" not in ir:
        errors.append("Missing 'items' field")
        return ValidationResult(False, errors, warnings)

    items = ir["items"]
    if not isinstance(items, dict):
        errors.append("'items' must be a dict")
        return ValidationResult(False, errors, warnings)

    # b0 entry
    if "b0" not in items:
        errors.append("Missing mandatory 'b0' entry node")
    else:
        b0 = items["b0"]
        if b0.get("type") != "branch":
            errors.append("b0 must have type 'branch'")
        if "branchId" not in b0:
            errors.append("b0 must have 'branchId'")
        if "one" not in b0:
            errors.append("b0 must have 'one' pointer")

    # end node
    if "end" not in items:
        errors.append("Missing mandatory 'end' node")
    else:
        if items["end"].get("type") != "end":
            errors.append("'end' node must have type 'end'")

    # params is string
    if "params" in ir and not isinstance(ir["params"], str):
        errors.append("'params' must be a string, not array/other type")

    # Node consistency
    all_ids = set(items.keys())
    for nid, node in items.items():
        ntype = node.get("type")
        if ntype == "action":
            if "content" not in node:
                warnings.append(f"Node {nid}: action missing 'content'")
            target = node.get("one")
            if target and target not in all_ids:
                errors.append(f"Node {nid}: 'one' refs unknown node '{target}'")
        elif ntype == "question":
            if "content" not in node:
                warnings.append(f"Node {nid}: question missing 'content'")
            else:
                if not node["content"].endswith("?"):
                    warnings.append(f"Node {nid}: question content should end with '?'")
            for key in ("one", "two"):
                target = node.get(key)
                if not target:
                    errors.append(f"Node {nid}: question missing '{key}' pointer")
                elif target not in all_ids:
                    errors.append(f"Node {nid}: '{key}' refs unknown node '{target}'")
        elif ntype == "branch":
            target = node.get("one")
            if target and target not in all_ids:
                errors.append(f"Node {nid}: 'one' refs unknown node '{target}'")
        elif ntype != "end":
            warnings.append(f"Node {nid}: unknown type '{ntype}'")

    # Reachability from b0
    if "b0" in items:
        reachable = _reachable(items, "b0")
        for nid in all_ids:
            if nid not in reachable:
                warnings.append(f"Node {nid} is unreachable from b0")

    return ValidationResult(len(errors) == 0, errors, warnings)


def _reachable(items: dict, start: str) -> set[str]:
    visited: set[str] = set()
    stack = [start]
    while stack:
        nid = stack.pop()
        if nid in visited or nid not in items:
            continue
        visited.add(nid)
        node = items[nid]
        for key in ("one", "two"):
            target = node.get(key)
            if target:
                stack.append(target)
    return visited

File: services/architect-agent-flue/lib/ir-validator.ts
export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

export function validateIr(ir: any): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (!ir || typeof ir !== 'object' || Array.isArray(ir)) {
    return { valid: false, errors: ['IR must be a dict'], warnings };
  }

  // Required fields
  if (!('name' in ir)) {
    errors.push("Missing 'name' field");
  }
  if (!('items' in ir)) {
    errors.push("Missing 'items' field");
    return { valid: false, errors, warnings };
  }

  const items = ir.items;
  if (!items || typeof items !== 'object' || Array.isArray(items)) {
    errors.push("'items' must be a dict");
    return { valid: false, errors, warnings };
  }

  // b0 entry
  if (!('b0' in items)) {
    errors.push("Missing mandatory 'b0' entry node");
  } else {
    const b0 = items.b0;
    if (b0.type !== 'branch') {
      errors.push("b0 must have type 'branch'");
    }
    if (!('branchId' in b0)) {
      errors.push("b0 must have 'branchId'");
    }
    if (!('one' in b0)) {
      errors.push("b0 must have 'one' pointer");
    }
  }

  // end node
  if (!('end' in items)) {
    errors.push("Missing mandatory 'end' node");
  } else {
    if (items.end?.type !== 'end') {
      errors.push("'end' node must have type 'end'");
    }
  }

  // params is string
  if ('params' in ir && typeof ir.params !== 'string') {
    errors.push("'params' must be a string, not array/other type");
  }

  // Node consistency
  const allIds = new Set(Object.keys(items));
  for (const [nid, nodeAny] of Object.entries(items)) {
    const node = nodeAny as any;
    if (!node || typeof node !== 'object') continue;

    const ntype = node.type;
    if (ntype === 'action') {
      if (!('content' in node)) {
        warnings.push(`Node ${nid}: action missing 'content'`);
      }
      const target = node.one;
      if (target && !allIds.has(target)) {
        errors.push(`Node ${nid}: 'one' refs unknown node '${target}'`);
      }
    } else if (ntype === 'question') {
      if (!('content' in node)) {
        warnings.push(`Node ${nid}: question missing 'content'`);
      } else {
        if (typeof node.content === 'string' && !node.content.endsWith('?')) {
          warnings.push(`Node ${nid}: question content should end with '?'`);
        }
      }
      for (const key of ['one', 'two']) {
        const target = node[key];
        if (!target) {
          errors.push(`Node ${nid}: question missing '${key}' pointer`);
        } else if (!allIds.has(target)) {
          errors.push(`Node ${nid}: '${key}' refs unknown node '${target}'`);
        }
      }
    } else if (ntype === 'branch') {
      const target = node.one;
      if (target && !allIds.has(target)) {
        errors.push(`Node ${nid}: 'one' refs unknown node '${target}'`);
      }
    } else if (ntype !== 'end') {
      warnings.push(`Node ${nid}: unknown type '${ntype}'`);
    }
  }

  // Reachability from b0
  if ('b0' in items) {
    const reachable = getReachable(items, 'b0');
    for (const nid of allIds) {
      if (!reachable.has(nid)) {
        warnings.push(`Node ${nid} is unreachable from b0`);
      }
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings
  };
}

function getReachable(items: any, start: string): Set<string> {
  const visited = new Set<string>();
  const stack = [start];
  while (stack.length > 0) {
    const nid = stack.pop()!;
    if (visited.has(nid) || !(nid in items)) {
      continue;
    }
    visited.add(nid);
    const node = items[nid];
    if (node && typeof node === 'object') {
      for (const key of ['one', 'two']) {
        const target = node[key];
        if (target) {
          stack.push(target);
        }
      }
    }
  }
  return visited;
}
